lib/src/subscription.dart

```diff
@@ -0,0 +1,155 @@
+////////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+import 'dart:collection';
+
+import 'native/realm_core.dart';
+import 'realm_class.dart';
+import 'util.dart';
+
+class Subscription {
+  final SubscriptionHandle _handle;
+
+  Subscription._(this._handle);
+
+  @override
+  // ignore: hash_and_equals
+  bool operator ==(Object other) {
+    if (identical(this, other)) return true;
+    if (other is! Subscription) return false;
+    return realmCore.subscriptionEquals(this, other);
+  }
+}
+
+extension SubscriptionInternal on Subscription {
+  SubscriptionHandle get handle => _handle;
+}
+
+class _SubscriptionIterator implements Iterator<Subscription> {
+  int _index = -1;
+  final SubscriptionSet _subscriptions;
+
+  _SubscriptionIterator._(this._subscriptions);
+
+  @override
+  Subscription get current => _subscriptions.elementAt(_index);
+
+  @override
+  bool moveNext() => ++_index < _subscriptions.length;
+}
+
+enum SubscriptionSetState {
```

Comment:
Why not?

Reply:
Most are internal and don't convey meaningful information to users. I think both .net and Java collapse them to a handful of higher level states.

Reply:
This enum is exposed to users so if we underscore half the members, users will be getting weird values.

Reply:
I'm not sure I agree here. To me `uncommited` and `bootstrapping` are easy to understand. If the user don't care he can always, do the collapse him self. Why prevent him from getting the details?

Reply:
Because then we need to do a lot of explanation for things that offer little value to users. The only case where you'd get `uncommitted` is if you check the state inside an `update` block, which is kind of silly. The difference between `pending` and `bootstrapping` is also fairly subtle and inconsequential - there's no way to "cancel" a pending subscription set, so knowing whether the server has seen it or not does not bring any meaningful information to a user.

Reply:
Say I wan't to show to my user the state of sync, are we still trying to contact the server (Pending) are we in the process of downloading a lot stuff (Bootstrapping) or are we up and running (Completed)

Reply:
The user does an action, that implicitly changes the subscription set, and notices that data is still missing after 20s. Why is not valuable to distinguish between:
1. We haven't even been able to contact the server yet (Pending), and
2. We are fast at work fetching the data (Bootstrapping), but it takes time..

Reply:
I'm still not convinced, but I'm convinced I'm a minority 😄 .. I will adopt @blagoev's suggestion, at least until the improved enums lands in dart 2.17